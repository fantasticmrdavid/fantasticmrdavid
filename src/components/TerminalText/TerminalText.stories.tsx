import React from "react";
import { TerminalText } from "./TerminalText";
import { Meta } from "@storybook/react";

export default {
  title: "TerminalText",
  component: TerminalText
} as Meta

export const Default = () => <TerminalText blinker>The quick brown fox jumps over the lazy dog.</TerminalText>
