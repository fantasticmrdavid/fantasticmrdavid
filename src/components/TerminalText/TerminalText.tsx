import React, { PureComponent, ReactElement } from 'react';
import { Blinker } from 'components/Blinker/Blinker';

const LETTER_INTERVAL = 30;
const COMPLETE_DELAY_INTERVAL = 2000;

interface TerminalTextProps {
  children?: string | ReactElement| undefined,
  blinker: boolean,
  onComplete?: (...args: any[]) => any,
  speed?: number,
}

type State = {
  text: ReactElement,
  index: number,
  letters: string[],
  currentLine: string,
  complete: boolean,
  paused: boolean,
};

export class TerminalText extends PureComponent<TerminalTextProps, State> {
  boundAddChar: () => void;

  boundPauseTypewriter: () => void;

  boundStartTypewriter: () => void;

  typewriter: number | undefined;

  constructor(props: TerminalTextProps) {
    super(props);
    const { children } = props;
    this.state = {
      text: <div />,
      index: 0,
      letters: children && typeof children === 'string' ? children.split('') : [],
      currentLine: '',
      complete: false,
      paused: false,
    };

    this.boundAddChar = this.addChar.bind(this);
    this.boundPauseTypewriter = this.pauseTypewriter.bind(this);
    this.boundStartTypewriter = this.startTypewriter.bind(this);
    this.boundStartTypewriter();
  }

  componentWillUnmount() {
    this.boundPauseTypewriter();
  }

  addChar() {
    const {
      complete,
      currentLine,
      index,
      letters,
      text,
    } = this.state;

    const {
      onComplete,
    } = this.props;

    if (complete) {
      this.boundPauseTypewriter();
      setTimeout(() => {
        if (!!onComplete) onComplete();
        return true;
      }, COMPLETE_DELAY_INTERVAL);
    }

    const nextChar = letters[index];

    if (index < letters.length) {
      this.setState({
        text: nextChar === '\n' ? <>{text}{currentLine}<br /></> : <>{text}</>,
        currentLine: nextChar === '\n' ? '' : `${currentLine}${nextChar}`,
        index: index + 1,
      });
    } else {
      this.setState({ complete: true });
    }
  }

  pauseTypewriter() {
    clearInterval(this.typewriter);
    this.setState({ paused: true });
  }

  startTypewriter() {
    const { speed } = this.props;
    const { paused } = this.state;
    this.typewriter = window.setInterval(() => {
      this.boundAddChar();
    }, speed || LETTER_INTERVAL);
    if (paused) this.setState({ paused: false });
  }

  render() {
    const {
      currentLine,
      text,
    } = this.state;
    const { blinker } = this.props;

    return (
      <>
        {text}{currentLine}
        {blinker && <Blinker />}
      </>
    );
  }
}

