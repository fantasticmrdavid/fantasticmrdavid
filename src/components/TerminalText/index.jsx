import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import Blinker from 'components/Blinker';

const LETTER_INTERVAL = 30;
const COMPLETE_DELAY_INTERVAL = 2000;

class TerminalText extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      text: <Fragment />,
      index: 0,
      letters: props.children.split(''),
      currentLine: '',
      complete: false,
      paused: false,
    };

    this.boundAddChar = this.addChar.bind(this);
    this.boundPauseTypewriter = this.pauseTypewriter.bind(this);
    this.boundStartTypewriter = this.startTypewriter.bind(this);
    this.boundStartTypewriter();
  }

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this.boundPauseTypewriter();
    this._isMounted = false;
  }

  addChar() {
    const {
      complete,
      currentLine,
      index,
      letters,
      text,
    } = this.state;

    const {
      onComplete,
    } = this.props;

    if (complete) {
      this.boundPauseTypewriter();
      setTimeout(() => {
        if (!!onComplete) onComplete();
        return true;
      }, COMPLETE_DELAY_INTERVAL);
    }

    const nextChar = letters[index];

    return this.setState(index < letters.length
      ? {
        text: nextChar === '\n' ? <Fragment>{text}{currentLine}<br /></Fragment> : <Fragment>{text}</Fragment>,
        currentLine: nextChar === '\n' ? '' : `${currentLine}${nextChar}`,
        index: index + 1,
      } : {
        complete: true,
      });
  }

  pauseTypewriter() {
    clearInterval(this.typewriter);
    this.setState({ paused: true });
  }

  startTypewriter() {
    const { speed } = this.props;
    const { paused } = this.state;
    this.typewriter = setInterval(() => {
      this.boundAddChar();
    }, speed || LETTER_INTERVAL);
    if (paused) this.setState({ paused: false });
  }

  render() {
    const {
      complete,
      currentLine,
      paused,
      text,
    } = this.state;
    const { blinker } = this.props;

    return (
      <Fragment>
        {text}{currentLine}
        {blinker && <Blinker solid={!complete && !paused} />}
      </Fragment>);
  }
}

TerminalText.propTypes = {
  children: PropTypes.node.isRequired,
  blinker: PropTypes.bool,
  onComplete: PropTypes.func,
  speed: PropTypes.number,
};

export default TerminalText;
