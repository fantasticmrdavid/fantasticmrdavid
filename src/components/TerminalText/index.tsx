import React, { Fragment, PureComponent, ReactElement } from 'react';
import Blinker from 'components/Blinker';

const LETTER_INTERVAL = 30;
const COMPLETE_DELAY_INTERVAL = 2000;

interface Props {
  children?: string | undefined,
  blinker: boolean,
  onComplete: (...args: any[]) => any,
  speed: number,
}

type State = {
  text: ReactElement,
  index: number,
  letters: string[],
  currentLine: string,
  complete: boolean,
  paused: boolean,
}

class TerminalText extends PureComponent<Props, State> {
  boundAddChar: () => void;

  boundPauseTypewriter: () => void;

  boundStartTypewriter: () => void;

  typewriter: number | undefined;

  constructor(props: Props) {
    super(props);
    const { children } = props;
    this.state = {
      text: <Fragment />,
      index: 0,
      letters: children ? children.split('') : [],
      currentLine: '',
      complete: false,
      paused: false,
    };

    this.boundAddChar = this.addChar.bind(this);
    this.boundPauseTypewriter = this.pauseTypewriter.bind(this);
    this.boundStartTypewriter = this.startTypewriter.bind(this);
    this.boundStartTypewriter();
  }

  componentWillUnmount() {
    this.boundPauseTypewriter();
  }

  addChar() {
    const {
      complete,
      currentLine,
      index,
      letters,
      text,
    } = this.state;

    const {
      onComplete,
    } = this.props;

    if (complete) {
      this.boundPauseTypewriter();
      setTimeout(() => {
        if (!!onComplete) onComplete();
        return true;
      }, COMPLETE_DELAY_INTERVAL);
    }

    const nextChar = letters[index];

    if (index < letters.length) {
      this.setState({
        text: nextChar === '\n' ? <Fragment>{text}{currentLine}<br /></Fragment> : <Fragment>{text}</Fragment>,
        currentLine: nextChar === '\n' ? '' : `${currentLine}${nextChar}`,
        index: index + 1,
      });
    } else {
      this.setState({ complete: true });
    }
  }

  pauseTypewriter() {
    clearInterval(this.typewriter);
    this.setState({ paused: true });
  }

  startTypewriter() {
    const { speed } = this.props;
    const { paused } = this.state;
    this.typewriter = window.setInterval(() => {
      this.boundAddChar();
    }, speed || LETTER_INTERVAL);
    if (paused) this.setState({ paused: false });
  }

  render() {
    const {
      currentLine,
      text,
    } = this.state;
    const { blinker } = this.props;

    return (
      <Fragment>
        {text}{currentLine}
        {blinker && <Blinker />}
      </Fragment>
    );
  }
}

export default TerminalText;
