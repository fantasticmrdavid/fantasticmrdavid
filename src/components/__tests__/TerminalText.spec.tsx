import React from 'react';
import { render } from "@testing-library/react";
import TerminalText from 'components/TerminalText';

const testString = 'The quick brown fox jumps over the lazy dog...';

const props = {
  blinker: true,
  onComplete: jest.fn(),
  speed: 5,
};

describe('Component -> TerminalText', () => {
  it('should render the TerminalText with a string without throwing an error', () => {
    expect(render(
      <TerminalText {...props}>
        {testString}
      </TerminalText>,
    )).toMatchSnapshot();
  });
});

describe('Component -> TerminalText', () => {
  it('should render the TerminalText with a component without throwing an error', () => {
    expect(render(
      <TerminalText {...props}>
        <div>{testString}</div>
      </TerminalText>,
    )).toMatchSnapshot();
  });
});
