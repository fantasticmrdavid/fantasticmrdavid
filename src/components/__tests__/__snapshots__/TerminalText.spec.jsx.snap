// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> TerminalText should render the TerminalText with a component without throwing an error 1`] = `
<TerminalText
  blinker={true}
  onComplete={[MockFunction]}
  speed={5}
>
  <Memo()>
    <styles__Cursor>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "styles__Cursor-cbfjlf-0",
              "isStatic": true,
              "lastClassName": "kPyiRo",
              "rules": Array [
                "position:relative;display:inline-block;width:0.5rem;margin-left:0.15rem;animation:",
                Keyframes {
                  "id": "sc-keyframes-kFBPLh",
                  "inject": [Function],
                  "name": "kFBPLh",
                  "rules": Array [
                    "@-webkit-keyframes kFBPLh{0%{opacity:1.0;}50%{opacity:0.0;}100%{opacity:1.0;}}",
                    "@keyframes kFBPLh{0%{opacity:1.0;}50%{opacity:0.0;}100%{opacity:1.0;}}",
                  ],
                  "toString": [Function],
                },
                " 1s step-end infinite",
                ";border-bottom:2px solid ",
                "#aaa",
                ";",
              ],
            },
            "displayName": "styles__Cursor",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "styles__Cursor-cbfjlf-0",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="styles__Cursor-cbfjlf-0 kPyiRo"
        />
      </StyledComponent>
    </styles__Cursor>
  </Memo()>
</TerminalText>
`;

exports[`Component -> TerminalText should render the TerminalText with a string without throwing an error 1`] = `
<TerminalText
  blinker={true}
  onComplete={[MockFunction]}
  speed={5}
>
  <Memo()>
    <styles__Cursor>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "styles__Cursor-cbfjlf-0",
              "isStatic": true,
              "lastClassName": "kPyiRo",
              "rules": Array [
                "position:relative;display:inline-block;width:0.5rem;margin-left:0.15rem;animation:",
                Keyframes {
                  "id": "sc-keyframes-kFBPLh",
                  "inject": [Function],
                  "name": "kFBPLh",
                  "rules": Array [
                    "@-webkit-keyframes kFBPLh{0%{opacity:1.0;}50%{opacity:0.0;}100%{opacity:1.0;}}",
                    "@keyframes kFBPLh{0%{opacity:1.0;}50%{opacity:0.0;}100%{opacity:1.0;}}",
                  ],
                  "toString": [Function],
                },
                " 1s step-end infinite",
                ";border-bottom:2px solid ",
                "#aaa",
                ";",
              ],
            },
            "displayName": "styles__Cursor",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "styles__Cursor-cbfjlf-0",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="styles__Cursor-cbfjlf-0 kPyiRo"
        />
      </StyledComponent>
    </styles__Cursor>
  </Memo()>
</TerminalText>
`;
